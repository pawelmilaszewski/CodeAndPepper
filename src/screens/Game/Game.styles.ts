import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  main: {
    flex: 1,
    alignItems: 'center',
  },
  header: {
    flexDirection: 'row',
    width: '100%',
    justifyContent: 'space-evenly',
  },
  endButton: {
    marginBottom: 54,
    padding: 16,
  },
  endButtonText: {
    fontSize: 20,
    lineHeight: 24,
    fontWeight: 'bold',
  },
  container: {
    flex: 1,
    marginTop: 32,
    alignItems: 'center',
  },
  card: {
    height: 150,
    width: 250,
    margin: 16,
  },
  cardScore: {
    fontSize: 16,
  },
  winner: {
    fontSize: 20,
    marginVertical: 8,
  },
});
