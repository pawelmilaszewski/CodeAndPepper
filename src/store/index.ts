import {createStore} from 'redux';
import reducer from './main/reducer';

export type RootState = ReturnType<typeof reducer>;

export const store = createStore(reducer);
