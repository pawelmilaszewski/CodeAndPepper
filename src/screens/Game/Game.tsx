/* eslint-disable react-native/no-inline-styles */
import React, {useState, useEffect, useCallback, useMemo} from 'react';
import {View, Text, TouchableOpacity} from 'react-native';
import {useSelector, useDispatch} from 'react-redux';
import {useNavigation} from '@react-navigation/native';
import {StackNavigationProp} from '@react-navigation/stack';
import {Card, Title, Paragraph} from 'react-native-paper';
import axios from 'axios';

import {RootState} from '../../store';
import {
  increaseBestTopScore,
  increaseBestBottomScore,
} from '../../store/main/actions';
import {MainParamList} from '../../navigation/Navigator';
import {styles} from './Game.styles';
import {Colors} from '../../utils/colors';
import {People, Starship} from '../../utils/types';
import {CustomHeaderButton, Loader} from '../../components';

type ScreenProp = StackNavigationProp<MainParamList>;

type Item = {
  name: string;
  crew: number;
  mass: number;
};

const Game = () => {
  const navigation = useNavigation<ScreenProp>();
  const dispatch = useDispatch();
  const bestTopScore = useSelector((state: RootState) => state.bestTopScore);
  const bestBottomScore = useSelector(
    (state: RootState) => state.bestBottomScore,
  );
  const [loading, setLoading] = useState(false);
  const [isPeople, setIsPeople] = useState(true);
  const [topCard, setTopCard] = useState<Item>({} as Item);
  const [bottomCard, setBottomCard] = useState<Item>({} as Item);
  const [dataPeople, setDataPeople] = useState<People[]>([]);
  const [dataStarships, setDataStarships] = useState<Starship[]>([]);
  const goBack = () => navigation.goBack();

  const getCardDetails = useCallback(
    (setDetails: (value: Item) => void, data: Starship[] | People[]) => {
      let index = Math.floor(Math.random() * data.length);
      let crew = 0;
      let mass = 0;
      if (isPeople) {
        let temp = data[index] as People;
        mass = parseInt(temp?.mass, 10);
        setDetails({
          name: temp?.name,
          mass,
          crew,
        });
      } else {
        let temp = data[index] as Starship;
        crew = temp?.crew.includes('-')
          ? parseInt(temp?.crew.split('-')[1], 10)
          : temp?.crew.includes(',')
          ? parseInt(temp?.crew.split(',')[1], 10)
          : parseInt(temp?.crew, 10);
        setDetails({
          name: temp?.name,
          mass,
          crew,
        });
      }
    },
    [isPeople],
  );

  const getData = useCallback(async () => {
    setLoading(true);
    await axios
      .get(`https://swapi.dev/api/${isPeople ? 'people' : 'starships'}/`)
      .then(resp => {
        if (isPeople) {
          setDataPeople(resp.data.results);
          setDataStarships([]);
          getCardDetails(setTopCard, resp.data.results);
          getCardDetails(setBottomCard, resp.data.results);
        } else {
          setDataStarships(resp.data.results);
          setDataPeople([]);
          getCardDetails(setTopCard, resp.data.results);
          getCardDetails(setBottomCard, resp.data.results);
        }
        setLoading(false);
      })
      .catch(err => {
        setLoading(false);
        console.warn(err);
      });
  }, [isPeople, getCardDetails]);

  const play = () => {
    getCardDetails(setTopCard, isPeople ? dataPeople : dataStarships);
    getCardDetails(setBottomCard, isPeople ? dataPeople : dataStarships);
  };

  const increaseTop = useCallback(
    () => dispatch(increaseBestTopScore()),
    [dispatch],
  );
  const increaseBottom = useCallback(
    () => dispatch(increaseBestBottomScore()),
    [dispatch],
  );

  const getWinner = useMemo(() => {
    if (topCard.mass > bottomCard.mass || topCard.crew > bottomCard.crew) {
      increaseTop();
      return topCard.name;
    } else if (
      topCard.mass < bottomCard.mass ||
      topCard.crew < bottomCard.crew
    ) {
      increaseBottom();
      return bottomCard.name;
    } else {
      return 'Nobody';
    }
  }, [topCard, bottomCard, increaseBottom, increaseTop]);

  useEffect(() => {
    getData();
  }, [getData, isPeople]);

  const renderCard = useCallback(
    (card: Item) => (
      <Card
        style={[
          styles.card,
          {
            backgroundColor:
              getWinner === card.name
                ? Colors.GREEN
                : getWinner === 'Nobody'
                ? Colors.GRAY
                : Colors.RED,
          },
        ]}>
        <Card.Content>
          <Title>Name: {card.name}</Title>
          <Paragraph>
            {isPeople ? `Mass: ${card.mass}` : `Crew: ${card.crew}`}
          </Paragraph>
        </Card.Content>
      </Card>
    ),
    [isPeople, getWinner],
  );

  return (
    <View style={styles.main}>
      {<Loader loading={loading} />}
      <View style={styles.header}>
        <CustomHeaderButton
          text={'People'}
          onPress={() => setIsPeople(true)}
          selected={isPeople}
        />
        <CustomHeaderButton
          text={'Starships'}
          onPress={() => setIsPeople(false)}
          selected={!isPeople}
        />
      </View>
      <View style={styles.container}>
        <Text style={styles.winner} testID={'winner_text'}>
          {getWinner} wins!
        </Text>
        <Text style={styles.cardScore} testID={'top_card_text'}>
          Top card: {bestTopScore} wins
        </Text>
        {renderCard(topCard)}
        {renderCard(bottomCard)}
        <Text style={styles.cardScore} testID={'bottom_card_text'}>
          Bottom card: {bestBottomScore} wins
        </Text>
      </View>
      <View>
        <TouchableOpacity
          testID={'play_button'}
          onPress={play}
          style={[styles.endButton, {marginBottom: 0}]}>
          <Text style={styles.endButtonText}>Play Again</Text>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={goBack}
          style={styles.endButton}
          testID={'end_button'}>
          <Text style={styles.endButtonText}>End Game</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default Game;
