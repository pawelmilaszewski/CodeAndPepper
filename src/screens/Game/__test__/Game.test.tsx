import React from 'react';
import Game from '../Game';
import {render} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {store} from '../../../store';

const MockGameComponent = () => (
  <Provider store={store}>
    <Game />
  </Provider>
);

test('render game', () => {
  const tree = render(<MockGameComponent />);

  const {toJSON, getByTestId} = tree;

  expect(getByTestId('winner_text')).toBeTruthy();
  expect(getByTestId('top_card_text')).toBeTruthy();
  expect(getByTestId('bottom_card_text')).toBeTruthy();
  expect(getByTestId('play_button')).toBeTruthy();
  expect(getByTestId('end_button')).toBeTruthy();

  expect(toJSON()).toMatchSnapshot();
});
