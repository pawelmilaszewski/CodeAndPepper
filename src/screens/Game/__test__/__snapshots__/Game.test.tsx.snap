// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render game 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#00000075",
        "bottom": 0,
        "flex": 1,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 10,
      }
    }
    testID="loader"
  >
    <ActivityIndicator
      animating={true}
      color="#000000"
      size="large"
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0000FF",
          "borderRadius": 10,
          "height": 54,
          "justifyContent": "center",
          "marginTop": 32,
          "opacity": 1,
          "width": 124,
        }
      }
      testID="button"
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 20,
            "letterSpacing": 1,
            "lineHeight": 24,
          }
        }
        testID="text"
      >
        People
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0000FF50",
          "borderRadius": 10,
          "height": 54,
          "justifyContent": "center",
          "marginTop": 32,
          "opacity": 1,
          "width": 124,
        }
      }
      testID="button"
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 20,
            "letterSpacing": 1,
            "lineHeight": 24,
          }
        }
        testID="text"
      >
        Starships
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "marginTop": 32,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "marginVertical": 8,
        }
      }
      testID="winner_text"
    >
      Nobody
       wins!
    </Text>
    <Text
      style={
        Object {
          "fontSize": 16,
        }
      }
      testID="top_card_text"
    >
      Top card: 
      0
       wins
    </Text>
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#808080",
          "borderColor": "rgba(0, 0, 0, 0.12)",
          "borderRadius": 4,
          "elevation": 1,
          "height": 150,
          "margin": 16,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.5,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 0.75,
          "width": 250,
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "paddingHorizontal": 16,
              },
              Object {
                "paddingVertical": 16,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Name: 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Mass: undefined
          </Text>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#808080",
          "borderColor": "rgba(0, 0, 0, 0.12)",
          "borderRadius": 4,
          "elevation": 1,
          "height": 150,
          "margin": 16,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.5,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 0.75,
          "width": 250,
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "paddingHorizontal": 16,
              },
              Object {
                "paddingVertical": 16,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Name: 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Mass: undefined
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "fontSize": 16,
        }
      }
      testID="bottom_card_text"
    >
      Bottom card: 
      0
       wins
    </Text>
  </View>
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginBottom": 0,
          "opacity": 1,
          "padding": 16,
        }
      }
      testID="play_button"
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "bold",
            "lineHeight": 24,
          }
        }
      >
        Play Again
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginBottom": 54,
          "opacity": 1,
          "padding": 16,
        }
      }
      testID="end_button"
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "bold",
            "lineHeight": 24,
          }
        }
      >
        End Game
      </Text>
    </View>
  </View>
</View>
`;
